compiler/rustc_builtin_macros/src/assert/context.rs

```diff
@@ -237,6 +237,9 @@ impl<'cx, 'a> Context<'cx, 'a> {
                 self.manage_cond_expr(prefix);
                 self.manage_cond_expr(suffix);
             }
+            ExprKind::Matches(expr, _, _) => {
+                self.manage_cond_expr(expr);
+            }
```

Comment:
@est31 Someway, somehow it is necessary to have an indicator pointing to the scrutinee expression of `matches!` in order to capture its content. The use of regular declarative macros makes `matches!` **in this context** a opaque `ExprKind::MacCall` which can't be captured without further processing. In regards to macro 2.0, I personally don't know how it help here.

Reply:
Thanks @est31

Hum... Looks like a way forward is unclear, three different approaches were already presented (including adhoc processing inside `assert` as you commented) without much apparent success.

To avoid further back-and-forth, I think it is better to wait for @petrochenkov. Personally, I am happy with any strategy as long as `matches!` are expanded :)